### PR TITLE
WIP: Rewrite of the section on testing

### PR DESCRIPTION
**Pull request by NorfairKing**
Based on the discussion in #184

### CODE REVIEW

**Comment by NorfairKing** [comment]
CC @Gabriel439 , it doesn't look like I can add you as a reviewer.

Also cc @silky who had an opinion as well.

File: sotu.md [hunk 1]
```diff
@@ -1193,61 +1193,49 @@ overload, such as:
...
+standard which all other property-based testing implementations are measured
+against.  The reason property testing works so smoothly in Haskell is due to
+Haskell's strong static types and type class system.  The strong static types
+allow for automatic generation of what would otherwise be boiler plate;
+generators, shrinking, general properties.  The type class system simplifies
+how tests are defined.  Shrinking means that any failing test result can be
 automatically minimized by rerunning the check on smaller and smaller inputs
-until `QuickCheck` identifies the corner case that triggers the failure.
-
-Mocking is another area where Haskell shines because you can overload almost
```
Comment (NorfairKing) [inline sotu.md:1211]: I removed the section on mocking as it is a Bad Idea (but I recognise that this is an opinionated position so I'm happy to put it back).
Instead I added a bit more about integration testing.

File: sotu.md [hunk 2]
```diff
@@ -1193,61 +1193,49 @@ overload, such as:
...
-
-Haskell also supports most testing functionality that you expect from other
+Haskell supports all testing functionality that you expect from other
 languages, including:
 
 * standard package interfaces for testing
 * unit testing libraries
 * test result summaries and visualization
+* golden tests
+* integration tests with standard integrations like databases, web servers, ...
```
Comment (NorfairKing) [inline sotu.md:1220]: I can list the integrations that are available if appopriate:

- sqlite
- postgres
- mongo
- redis
- rabbitmq
- wai
- yesod
- servant

File: sotu.md [hunk 3]
```diff
@@ -1193,61 +1193,49 @@ overload, such as:
...
 
-**Rating:** Mature
-
-There are a few places where Haskell is the clear leader among all languages:
+**Rating:** Best in Class
 
-* property-based testing
-* mocking / dependency injection
+In terms of testing, Haskell is a clear leader.
+It has both more practical, more advanced and more advanced test automation than other languages.
```
Comment (jhrcek) [inline sotu.md]: "more advanced and more advanced" ?

Comment (NorfairKing) [reply sotu.md]: LOL, Will fix.

File: sotu.md (hunk 1 above)
Comment (ivanopagano) [reply sotu.md:1211]: You might want at least to add mocking as an available choice in the list of expected functionalities, outlined right after this paragraph

File: sotu.md [hunk 4]
```diff
@@ -1193,61 +1193,50 @@ overload, such as:
...
 * unit testing libraries
 * test result summaries and visualization
+* golden tests
+* mocking
+* integration tests with standard integrations like databases, web servers, ...
+
+On top of that, Haskell also supports more advanced testing features that are not common elsewhere:
+
+* Flakiness diagnostics
+* Deterministic randomness
```
Comment (freizl) [inline sotu.md:1225]: Would you help to elaborate what are those advanced testing features like how they has been implemented and/or what library does it? Thanks!

**Comment by nponeccop** [comment]
> In terms of testing, Haskell is a clear leader. It has both more practical, more robust and more advanced test automation than other languages.
> 
> Haskell's property testing, which started with QuickCheck, is the gold standard which all other property-based testing implementations are measured against. The reason property testing works so smoothly in Haskell is due to Haskell's strong static types and type class system. 
> 

What about this?

> Automated tests in Haskell are best in class.
> 
> In particular, property testing works better in Haskell than in other languages, thanks to Haskell's type and effect system, notably type classes. It all started with the QuickCheck library, which remains the gold standard against which all other property-based testing implementations are measured.

I wonder if we can support the claims about practicality, robustness, and being advanced. Otherwise, they are empty marketing words that are better replaced with a more neutral "best in class".

File: sotu.md (hunk 4 above)
Comment (NorfairKing) [reply sotu.md:1225]: https://github.com/NorfairKing/sydtest The readme here should answer these questions, please ping me if it doesn't.

**Comment by NorfairKing** [comment]
> > In terms of testing, Haskell is a clear leader. It has both more practical, more robust and more advanced test automation than other languages.
> > Haskell's property testing, which started with QuickCheck, is the gold standard which all other property-based testing implementations are measured against. The reason property testing works so smoothly in Haskell is due to Haskell's strong static types and type class system.
> 
> What about this?
> 
> > Automated tests in Haskell are best in class.
> > In particular, property testing works better in Haskell than in other languages, thanks to Haskell's type and effect system, notably type classes. It all started with the QuickCheck library, which remains the gold standard against which all other property-based testing implementations are measured.
> 
> I wonder if we can support the claims about practicality, robustness, and being advanced. Otherwise, they are empty marketing words that are better replaced with a more neutral "best in class".

I welcome you to challenge these claims because I don't know what I could say that would convince you.

File: sotu.md [hunk 5]
```diff
@@ -1193,61 +1193,50 @@ overload, such as:
...
+* Monadic test suite definition
+
 
 **Notable libraries:**
 
-* [`QuickCheck`](https://hackage.haskell.org/package/QuickCheck) - property-based testing
+- [`validity`](https://hackage.haskell.org/package/validity) - Property testing with smarter generators, Free generators, free shrinking, cheap properties
+- [`sydtest`](https://hackage.haskell.org/package/sydtest) - A modern testing framework for Haskell with good defaults and advanced testing features. It also has companion libraries for all major integration testing.
```
Comment (sjakobi) [inline sotu.md:1232]: `sydtest` appears to have a very restrictive [license](http://hackage.haskell.org/package/sydtest-0.1.0.0/src/LICENSE.md). Doesn't that make it unfit for recommendation in this guide? At the very least, a disclaimer should be added.

**Comment by nponeccop** [comment]
> I welcome you to challenge these claims because I don't know what I could say that would convince you.

Challenging is easy, but it's not my point. My point is that if you claim it's advanced you should explain how it's advanced to someone who doesn't know Haskell.

My 2 challenging points are about Haskell vs Typescript testing:

- I can easily mock filesystem calls in TypeScript using `memfs` and `jest.mock`. It isn't possible at all in Haskell thanks to its static nature and weak tooling.
- I can run single tests within a WebStorm IDE with Typescript, including under debugger. Does HIE support that yet? 

So you don't merely need to explain why and how it's advanced. You should explain why and how it's advanced despite having these obvious flaws for anyone who ever did testing in an industrial setting as opposed to some Linux command line fanboys.

Note that I'm not actually claiming it's not advanced, but it's the type of criticism we are going to receive. If there are short arguments in favor of Haskell I'd like to see them in place of mere claims of superiority

File: sotu.md (hunk 5 above)
Comment (NorfairKing) [reply sotu.md:1232]: `sydtest` is licenced such that I can maintain it in a sustainable way. This is not a limitation of Haskell the language and should not be a consideration here IMO

**Comment by NorfairKing** [comment]
> > I welcome you to challenge these claims because I don't know what I could say that would convince you.
> 
> Challenging is easy, but it's not my point. My point is that if you claim it's advanced you should explain how it's advanced to someone who doesn't know Haskell.
> 

It's right here:

```
On top of that, Haskell also supports more advanced testing features that are not common elsewhere:

* Flakiness diagnostics
* Deterministic randomness
* Monadic test suite definition
```
or is that not what you meant?


> My 2 challenging points are about Haskell vs Typescript testing:
> 
>     * I can easily mock filesystem calls in TypeScript using `memfs` and `jest.mock`. It isn't possible at all in Haskell thanks to its static nature and weak tooling.

You can 'mock' any syscalls using  https://github.com/nh2/hatrace (eventhough it's a bad idea to use mocking) and use something like free monads if that's too low-level for you.

>     * I can run single tests within a WebStorm IDE with Typescript, including under debugger. Does HIE support that yet?

I don't know about HIE but you can certainly run a single test and run `strace` or `valgrind` on the execution.
(Line-based debugging doesn't make sense in a lazy language so I'm not sure how that compares.)
 

> So you don't merely need to explain why and how it's advanced. You should explain why and how it's advanced despite having these obvious flaws for anyone who ever did testing in an industrial setting as opposed to some Linux command line fanboys.

I don't think these flaws are obvious, nor do I believe they exist.
I _am_ doing testing in an industrial setting and testing in haskell has been way ahead of anything else I've worked with.

> Note that I'm not actually claiming it's not advanced, but it's the type of criticism we are going to receive. If there are short arguments in favor of Haskell I'd like to see them in place of mere claims of superiority

That's certainly fair, but I'm not sure what such a "short argument" would look like.

**Comment by Gabriella439** [comment]
@NorfairKing: I think what @nponeccop means about `memfs` is that he's looking for a Haskell package that would actually implement an in-memory filesystem.  A free monad interface would allow you to switch out a real filesystem with an in-memory filesystem but you still need to actually implement the in-memory filesystem.  In other words, he's looking for an off-the-shelf package that can be used as the pure backend of a free monad.

**Comment by NorfairKing** [comment]
@Gabriel439 here you go: https://github.com/NorfairKing/dirforest/

**Comment by Gabriella439** [comment]
Based on this discussion, I think the main thing I'm looking for in order for this to be convincing is that there needs to be:

* a clear document (preferably a book) on how to do testing in Haskell with several end-to-end examples for common use cases
* a set of endorsed packages for this purpose that are available on Hackage (preferably Stackage, but not Github-only packages)

Like @sjakobi, I also think that a non-free license for any of these endorsed packages (e.g. `sydtest`) is problematic.  I get that you're trying to make it sustainable, which is fine and you're within your right to do so, but that goal is in conflict with the goal of making Haskell best in class for testing.  The other programming languages that we're competing with for the title of best in testing do not have the limitation of needing to pay for core testing packages.

**Comment by NorfairKing** [comment]
>    a clear document (preferably a book) on how to do testing in Haskell with several end-to-end examples for common use cases

The sydtest docs are full of them, but I don't have the time to write a book.
  
>  a set of endorsed packages for this purpose that are available on Hackage (preferably Stackage, but not Github-only packages)

The sydtest package is on hackage, not all of them yet but they'll get there.


> Like @sjakobi, I also think that a non-free license for any of these endorsed packages (e.g. sydtest) is problematic. I get that you're trying to make it sustainable, which is fine and you're within your right to do so, but that goal is in conflict with the goal of making Haskell best in class for testing. The other programming languages that we're competing with for the title of best in testing do not have the limitation of needing to pay for core testing packages.

I don't like how the sydtest license is being misrepresented here, but I get your point.
I think that this conversation has this gives me enough information to close the issue.
Thank you for your consideration and the time people took to review this.
